pkg/rc/replication_controller.go

```diff
@@ -633,81 +661,267 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, node types
-	if len(ineligibleCurrent) < 1 {
-		return util.Errorf("Need at least one ineligible node to transfer from")
+func (rc *replicationController) updateAllocations(ineligible []types.NodeName) (types.NodeName, error) {
+	if len(ineligible) < 1 {
+		return "", util.Errorf("Need at least one ineligible node to transfer from, had 0")
 	}
 
-	status, queryMeta, statusErr := rc.rcStatusStore.Get(rc.ID())
-	if statusErr != nil && !statusstore.IsNoStatus(statusErr) {
-		return util.Errorf("Unable to get rc status: %s", statusErr)
-	}
+	rc.mu.Lock()
+	man := rc.Manifest
+	sel := rc.NodeSelector
+	rc.mu.Unlock()
 
-	if status.NodeTransfer != nil {
-		// Node transfer already in progress
-		return nil
+	oldNode := ineligible[0]
+	err := rc.scheduler.DeallocateNodes(sel, []types.NodeName{oldNode})
+	if err != nil {
+		return "", util.Errorf("Could not deallocate from %s: %s", oldNode, err)
 	}
 
 	nodesRequested := 1 // We only support one node transfer at a time right now
-	rc.mu.Lock()
-	newNodes, err := rc.scheduler.AllocateNodes(rc.Manifest, rc.NodeSelector, nodesRequested)
-	rc.mu.Unlock()
+	newNodes, err := rc.scheduler.AllocateNodes(man, sel, nodesRequested)
 	if err != nil || len(newNodes) < 1 {
 		errMsg := fmt.Sprintf("Unable to allocate nodes over grpc: %s", err)
 		err := rc.alerter.Alert(rc.alertInfo(errMsg), alerting.LowUrgency)
 		if err != nil {
 			rc.logger.WithError(err).Errorln("Unable to send alert")
 		}
 
-		return util.Errorf(errMsg)
+		return "", util.Errorf(errMsg)
 	}
 
-	var modifyIndex uint64
-	if statusErr != nil && statusstore.IsNoStatus(statusErr) {
-		modifyIndex = 0
-	} else {
-		modifyIndex = queryMeta.LastIndex
-	}
+	newNode := newNodes[0]
 
-	status.NodeTransfer = &rcstatus.NodeTransfer{
-		// Again, we only support one node transfer at a time so just take the first old and new
-		OldNode: ineligibleCurrent[0],
-		NewNode: newNodes[0],
+	status := rcstatus.Status{
+		NodeTransfer: &rcstatus.NodeTransfer{
+			OldNode: oldNode,
+			NewNode: newNode,
+		},
 	}
+
 	writeCtx, writeCancel := transaction.New(context.Background())
 	defer writeCancel()
-	err = rc.rcStatusStore.CASTxn(writeCtx, rc.ID(), modifyIndex, status)
+	err = rc.rcStatusStore.CASTxn(writeCtx, rc.ID(), 0, status)
 	if err != nil {
-		return util.Errorf("Could not write new node to store: %s", err)
+		return "", util.Errorf("Could not write new node to store: %s", err)
 	}
 
 	err = transaction.MustCommit(writeCtx, rc.txner)
 	if err != nil {
-		return util.Errorf("Could not commit CASTxn: %s", err)
+		return "", util.Errorf("Could not commit CASTxn: %s", err)
+	}
+
+	return newNode, nil
+}
+
+func (rc *replicationController) scheduleWithSession(newNode types.NodeName) error {
+	rc.logger.NoFields().Infof("Scheduling on %s", newNode)
+	rc.mu.Lock()
+	manifest := rc.Manifest
+	rc.mu.Unlock()
+
+	manifestBytes, err := manifest.Marshal()
+	if err != nil {
+		return err
+	}
+
+	key := path.Join(string(consul.INTENT_TREE), newNode.String(), manifest.ID().String())
```

Comment:
this string computation should probably be put in `pkg/store/consul` somewhere with all the safety checks around empty hostname and pod ID, etc